File: backend/app/backtest/sweep.py
"""Stability / sensitivity sweep (backend mirror of frontend/src/engine/sweep.ts).

Varies a single parameter (optionally split into a few series) and reports a
chosen metric across its range, so robustness can be judged as a plateau rather
than a single spike. Built on top of the grid engine.
"""
from __future__ import annotations

from typing import Literal

import pandas as pd
from pydantic import BaseModel

from ..sessions import DEFAULT_SESSIONS, Session
from ..strategies.follow_filters import parse_hhmm
from .engine import BacktestConfig, Strategy, make_runner
from .grid import LevelRange, NumRange, ToggleRange, expand_grid, GridSpec

SweepParam = Literal[
    "entry_delay", "entry_time", "entry_timeout", "invert_multiple", "invert_offset",
    "time_stop", "gap_window", "gap_sigma", "sl_value", "tp_value",
    "invert_sl_value", "invert_tp_value",
]
SweepMetric = Literal[
    "total_pnl", "return_dd", "profit_factor", "total_r", "win_rate", "expectancy", "trades"
]
SeriesBy = Literal["none", "direction", "session"]


class SweepSpec(BaseModel):
    param: SweepParam
    min: float
    max: float
    step: float
    series: SeriesBy = "none"
    metric: SweepMetric = "total_pnl"


class SweepRequest(BaseModel):
    base: BacktestConfig
    spec: SweepSpec


def _entry_hours_after_open(config: BacktestConfig) -> float:
    """A follow config's entry time as hours after the session open (0..24), the
    duration the entry_time sweep ranges over. Inverse of the grid's resolution."""
    if not config.entry_times:
        return 0.0
    m = parse_hhmm(config.entry_times[0])
    clock = (m / 60) if m is not None else 0.0
    s = DEFAULT_SESSIONS.get(config.session)
    open_h = (s.open_time.hour + s.open_time.minute / 60) if s else 9.5
    return (clock - open_h) % 24


def _base_entry_hour(base: BacktestConfig) -> float:
    """Neutral fixed value when entry_time isn't swept (then it's unused anyway)."""
    return _entry_hours_after_open(base)


def _fixed(v: float) -> NumRange:
    return NumRange(vary=False, fixed=v, min=v, max=v, step=1)


def _varied(s: SweepSpec) -> NumRange:
    return NumRange(vary=True, fixed=s.min, min=s.min, max=s.max, step=s.step)


def build_grid_spec(base: BacktestConfig, spec: SweepSpec) -> GridSpec:
    p = spec.param
    # follow_filters is follow-only, so the "direction" series collapses to follow.
    is_follow = base.strategy == Strategy.follow_filters
    return GridSpec(
        strategy=base.strategy,
        sessions=[s.name for s in DEFAULT_SESSIONS.values()] if spec.series == "session" else [base.session],
        directions=["fade", "follow"] if spec.series == "direction" and not is_follow else [base.direction],
        gap_window=_varied(spec) if p == "gap_window" else _fixed(base.gap_window),
        gap_sigma=_varied(spec) if p == "gap_sigma" else _fixed(base.gap_sigma),
        entry_offset_hours=_varied(spec) if p == "entry_delay" else _fixed(base.entry_offset_minutes / 60),
        entry_times=base.entry_times,
        # When not sweeping entry_time, leave it non-varying so the fixed
        # entry_times list above is used (its value is then irrelevant).
        entry_time=_varied(spec) if p == "entry_time" else _fixed(_base_entry_hour(base)),
        entry_timeout=_varied(spec) if p == "entry_timeout" else _fixed(base.entry_timeout_minutes / 60),
        invert=[base.invert_enabled],  # carry the base's inversion setting through the sweep
        invert_multiple=_varied(spec) if p == "invert_multiple" else _fixed(base.invert_gap_multiple),
        invert_offset_hours=_varied(spec) if p == "invert_offset" else _fixed(base.invert_entry_offset_minutes / 60),
        invert_custom_exits=base.invert_custom_exits or p in ("invert_sl_value", "invert_tp_value"),
        invert_sl=LevelRange(
            enabled=base.invert_stop_loss is not None or p == "invert_sl_value",
            mode=base.invert_stop_loss.mode if base.invert_stop_loss else "adr_multiple",
            **(_varied(spec) if p == "invert_sl_value" else _fixed(base.invert_stop_loss.value if base.invert_stop_loss else 0.5)).model_dump(),
        ),
        invert_tp=LevelRange(
            enabled=base.invert_take_profit is not None or p == "invert_tp_value",
            mode=base.invert_take_profit.mode if base.invert_take_profit else "adr_multiple",
            **(_varied(spec) if p == "invert_tp_value" else _fixed(base.invert_take_profit.value if base.invert_take_profit else 1.0)).model_dump(),
        ),
        time_stop=ToggleRange(
            enabled=base.time_stop_minutes is not None or p == "time_stop",
            **(_varied(spec) if p == "time_stop" else _fixed((base.time_stop_minutes or 1440) / 60)).model_dump(),
        ),
        sl=LevelRange(
            enabled=base.stop_loss is not None or p == "sl_value",
            mode=base.stop_loss.mode if base.stop_loss else "adr_multiple",
            **(_varied(spec) if p == "sl_value" else _fixed(base.stop_loss.value if base.stop_loss else 0.5)).model_dump(),
        ),
        tp=LevelRange(
            enabled=base.take_profit is not None or p == "tp_value",
            mode=base.take_profit.mode if base.take_profit else "adr_multiple",
            **(_varied(spec) if p == "tp_value" else _fixed(base.take_profit.value if base.take_profit else 1.0)).model_dump(),
        ),
        spread=base.spread,
    )


def extract_x(config: BacktestConfig, param: SweepParam) -> float:
    if param == "entry_delay":
        return config.entry_offset_minutes / 60
    if param == "entry_time":
        return _entry_hours_after_open(config)
    if param == "entry_timeout":
        return config.entry_timeout_minutes / 60
    if param == "invert_multiple":
        return config.invert_gap_multiple
    if param == "invert_offset":
        return config.invert_entry_offset_minutes / 60
    if param == "time_stop":
        return (config.time_stop_minutes or 0) / 60
    if param == "gap_window":
        return config.gap_window
    if param == "gap_sigma":
        return config.gap_sigma
    if param == "invert_sl_value":
        return config.invert_stop_loss.value if config.invert_stop_loss else 0
    if param == "invert_tp_value":
        return config.invert_take_profit.value if config.invert_take_profit else 0
    if param == "sl_value":
        return config.stop_loss.value if config.stop_loss else 0
    return config.take_profit.value if config.take_profit else 0


def get_metric(metrics: dict, metric: SweepMetric) -> float | None:
    if metric == "win_rate":
        return metrics["win_rate"] * 100
    if metric == "return_dd":
        dd = metrics["max_drawdown"]
        return metrics["total_pnl"] / dd if dd > 0 else None
    return metrics.get(metric)


def run_sweep(df_utc: pd.DataFrame, sessions: dict[str, Session], base: BacktestConfig, spec: SweepSpec) -> dict:
    configs = expand_grid(build_grid_spec(base, spec))
    run = make_runner(df_utc, sessions)  # memoizes signal-level work across the sweep
    lines: dict[str, list[dict]] = {}
    for config in configs:
        metrics = run(config)["metrics"]
        if spec.series == "direction":
            label = config.direction
        elif spec.series == "session":
            label = config.session
        else:
            label = "result"
        lines.setdefault(label, []).append(
            {"x": extract_x(config, spec.param), "y": get_metric(metrics, spec.metric)}
        )

    series = [
        {"label": label, "points": sorted(points, key=lambda p: p["x"])}
        for label, points in lines.items()
    ]
    return {"series": series, "param": spec.param, "metric": spec.metric}

File: frontend/src/engine/sweep.ts
// Stability / sensitivity sweep: vary one parameter (optionally split into a few
// series) and report a chosen metric across its range, so robustness can be
// eyeballed as a plateau vs. a single spike. Built on top of the grid engine.

import type { BacktestConfig, Bar } from "./types";
import { DEFAULT_SESSIONS } from "./sessions";
import { expandGrid, hhmmToHours, type GridSpec, type NumRange } from "./grid";
import { getSession } from "./sessions";
import { runBacktest } from "./backtest";
import { parseHHMM } from "./followFilters";

// The base config's first entry time as hours-after-open (the neutral fixed value
// when entry_time isn't the swept parameter). Falls back to 0.
function baseEntryHour(base: BacktestConfig): number {
  return entryHoursAfterOpen(base);
}

// A follow config's entry time expressed as hours after the session open (0..24),
// the duration the entry_time sweep ranges over. Inverse of the grid's resolution.
function entryHoursAfterOpen(config: BacktestConfig): number {
  const clock = (parseHHMM(config.entry_times[0] ?? "") ?? 0) / 60;
  const open = hhmmToHours(getSession(config.session).open_time);
  return ((clock - open) % 24 + 24) % 24;
}

export type SweepParam =
  | "entry_delay"
  | "entry_time"
  | "entry_timeout"
  | "time_stop"
  | "gap_window"
  | "gap_sigma"
  | "sl_value"
  | "tp_value";

export type SweepMetric =
  | "total_pnl"
  | "return_dd"
  | "profit_factor"
  | "total_r"
  | "win_rate"
  | "expectancy"
  | "trades";

export type SeriesBy = "none" | "direction" | "session";

export interface SweepSpec {
  param: SweepParam;
  min: number;
  max: number;
  step: number;
  series: SeriesBy;
  metric: SweepMetric;
}

export interface SweepSeries {
  label: string;
  points: { x: number; y: number | null }[];
}

export interface SweepResult {
  series: SweepSeries[];
  param: SweepParam;
  metric: SweepMetric;
}

export const PARAM_LABELS: Record<SweepParam, string> = {
  entry_delay: "Entry delay (h)",
  entry_time: "Entry: hrs after open",
  entry_timeout: "Wait timeout (h)",
  time_stop: "Time stop (h)",
  gap_window: "Gap window",
  gap_sigma: "Gap sigma",
  sl_value: "Stop loss",
  tp_value: "Take profit",
};

export const METRIC_LABELS: Record<SweepMetric, string> = {
  total_pnl: "Total P/L",
  return_dd: "Return / Max DD",
  profit_factor: "Profit factor",
  total_r: "Total R",
  win_rate: "Win rate %",
  expectancy: "Expectancy",
  trades: "Trades",
};

function fixed(value: number): NumRange {
  return { vary: false, fixed: value, min: value, max: value, step: 1 };
}

function varied(spec: SweepSpec): NumRange {
  return { vary: true, fixed: spec.min, min: spec.min, max: spec.max, step: spec.step };
}

// Translate the base config + sweep selection into a one-parameter grid.
function buildGridSpec(base: BacktestConfig, spec: SweepSpec): GridSpec {
  const p = spec.param;
  const isFollow = base.strategy === "follow_filters";
  return {
    strategy: base.strategy,
    sessions: spec.series === "session" ? DEFAULT_SESSIONS.map((s) => s.name) : [base.session],
    // follow_filters is follow-only, so the "direction" series collapses to follow.
    directions: spec.series === "direction" && !isFollow ? ["fade", "follow"] : [base.direction],
    gapWindow: p === "gap_window" ? varied(spec) : fixed(base.gap_window),
    gapSigma: p === "gap_sigma" ? varied(spec) : fixed(base.gap_sigma),
    entryOffsetHours:
      p === "entry_delay" ? varied(spec) : fixed(base.entry_offset_minutes / 60),
    entryTimes: base.entry_times,
    // When not sweeping entry_time, leave entryTime non-varying so the fixed
    // entryTimes list above is used (its value is then irrelevant).
    entryTime: p === "entry_time" ? varied(spec) : fixed(baseEntryHour(base)),
    entryTime2: fixed(0), // the stability sweep varies a single parameter only
    entryTimeout: p === "entry_timeout" ? varied(spec) : fixed(base.entry_timeout_minutes / 60),
    timeStop: {
      enabled: base.time_stop_minutes != null || p === "time_stop",
      ...(p === "time_stop" ? varied(spec) : fixed((base.time_stop_minutes ?? 1440) / 60)),
    },
    sl: {
      enabled: base.stop_loss != null || p === "sl_value",
      mode: base.stop_loss?.mode ?? "adr_multiple",
      ...(p === "sl_value" ? varied(spec) : fixed(base.stop_loss?.value ?? 0.5)),
    },
    tp: {
      enabled: base.take_profit != null || p === "tp_value",
      mode: base.take_profit?.mode ?? "adr_multiple",
      ...(p === "tp_value" ? varied(spec) : fixed(base.take_profit?.value ?? 1.0)),
    },
    spread: base.spread, // carry the base's spread through the sweep
    rankBy: "total_pnl",
  };
}

export function extractX(config: BacktestConfig, param: SweepParam): number {
  switch (param) {
    case "entry_delay":
      return config.entry_offset_minutes / 60;
    case "entry_time":
      return entryHoursAfterOpen(config);
    case "entry_timeout":
      return config.entry_timeout_minutes / 60;
    case "time_stop":
      return (config.time_stop_minutes ?? 0) / 60;
    case "gap_window":
      return config.gap_window;
    case "gap_sigma":
      return config.gap_sigma;
    case "sl_value":
      return config.stop_loss?.value ?? 0;
    case "tp_value":
      return config.take_profit?.value ?? 0;
  }
}

export function getMetric(
  m: {
    total_pnl: number;
    max_drawdown: number;
    profit_factor: number | null;
    total_r: number | null;
    win_rate: number;
    expectancy: number;
    trades: number;
  },
  metric: SweepMetric
): number | null {
  switch (metric) {
    case "total_pnl":
      return m.total_pnl;
    case "return_dd":
      // Plotting: leave undefined (no drawdown yet) as a gap in the line.
      return m.max_drawdown > 0 ? m.total_pnl / m.max_drawdown : null;
    case "profit_factor":
      return m.profit_factor;
    case "total_r":
      return m.total_r;
    case "win_rate":
      return m.win_rate * 100;
    case "expectancy":
      return m.expectancy;
    case "trades":
      return m.trades;
  }
}

function seriesLabel(config: BacktestConfig, by: SeriesBy): string {
  return by === "session" ? config.session : config.direction;
}

export function runSweep(bars: Bar[], base: BacktestConfig, spec: SweepSpec): SweepResult {
  const configs = expandGrid(buildGridSpec(base, spec));
  const lines = new Map<string, { x: number; y: number | null }[]>();

  for (const config of configs) {
    const metrics = runBacktest(bars, getSession(config.session), config).metrics;
    const label =
      spec.series === "none" ? METRIC_LABELS[spec.metric] : seriesLabel(config, spec.series);
    if (!lines.has(label)) lines.set(label, []);
    lines.get(label)!.push({ x: extractX(config, spec.param), y: getMetric(metrics, spec.metric) });
  }

  const series: SweepSeries[] = [...lines.entries()].map(([label, points]) => ({
    label,
    points: points.sort((a, b) => a.x - b.x),
  }));
  return { series, param: spec.param, metric: spec.metric };
}
